library/core/src/ptr/const_ptr.rs

```diff
@@ -704,6 +705,16 @@ impl<T: ?Sized> *const T {
     #[stable(feature = "pointer_byte_offsets", since = "1.75.0")]
     #[rustc_const_stable(feature = "const_pointer_byte_offsets", since = "1.75.0")]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        (mem::size_of_val_raw(self) != 0) &&
+        // Ensures subtracting `origin` from `self` doesn't overflow
+        (self.addr() as isize).checked_sub(origin.addr() as isize).is_some() &&
+        // Ensure both pointers are in the same allocation or are pointing to the same address
+        (self.addr() == origin.addr() ||
+            kani::mem::same_allocation(self as *const u8, origin as *const u8))
+    )]
+    // The result should equal the distance in terms of bytes
+    #[ensures(|result| *result == (self.addr() as isize - origin.addr() as isize))]
```

Comment:
Could you please add at least one harness that actually uses this post-condition and checks that it indeed is encoding the correct behaviour? For example, do a harness where `origin` and `self` are the same pointer so that the result must be 0, and another one that has `self` with a known constant offset from another pointer that is then used as `origin`.

Reply:
@tautschnig Are you looking for a harness like this:

```rust
#[kani::proof]
pub fn check_const_byte_offset_from_fixed_offset() {
    let arr: [u32; ARRAY_LEN] = kani::Arbitrary::any_array();
    const offset: isize = 25;
    let origin_ptr: *const u32 = arr.as_ptr();
    let self_ptr: *const u32 = unsafe { origin_ptr.byte_offset(offset) };
    let result: isize = unsafe { self_ptr.byte_offset_from(origin_ptr) };
    assert_eq!(result, offset);
    assert_eq!(result, (self.addr() as isize - origin.addr() as isize));
}
```

Reply:
Yes, thank you! But now that you spelled it out I am thinking we could do a tiny bit better and make the offset arbitrary (though still less than `ARRAY_LEN`)?

Reply:
Sure, I'll use `kani::any_where` to do that and update the PR.

Reply:
@tautschnig , I updated the PR with the required harnesses.